### ISSUE TITLE
error "Found unknown session xx"

### ISSUE THREAD

**Issue by ray3888** [orientechnologies/orientdb#6170]
Found a lot of following errors in server log at peak hours:
> Error executing request
com.orientechnologies.common.io.OIOException: Found unknown session 814
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.onBeforeRequest(ONetworkProtocolBinary.java:240)
        at com.orientechnologies.orient.server.network.protocol.binary.OBinaryNetworkProtocolAbstract.execute(OBinaryNetworkProtocolAbstract.java:214)
        at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:77)

 and some errors in client log (much less than server log):

> error=Error on executing command: sql.select from v where pk='150001984621'
com.orientechnologies.orient.core.exception.OStorageException: Error on executing command: sql.select from v where pk='150001984621'
        at com.orientechnologies.orient.client.remote.OStorageRemote.handleException(OStorageRemote.java:1841) ~[orientdb-client-2.1.7.jar:2.1.7]
        at com.orientechnologies.orient.client.remote.OStorageRemote.command(OStorageRemote.java:1241) ~[orientdb-client-2.1.7.jar:2.1.7]
        at com.orientechnologies.orient.client.remote.OStorageRemoteThread.command(OStorageRemoteThread.java:444) ~[orientdb-client-2.1.7.jar:2.1.7]
        at com.orientechnologies.orient.core.command.OCommandRequestTextAbstract.execute(OCommandRequestTextAbstract.java:67) ~[orientdb-core-2.1.7.jar:2.1.7]
        at com.tinkerpop.blueprints.impls.orient.OrientGraphCommand.execute(OrientGraphCommand.java:49) ~[orientdb-graphdb-2.1.7.jar:2.1.7]
Caused by: java.io.EOFException: null
        at java.io.DataInputStream.readByte(DataInputStream.java:267) ~[na:1.7.0_79]
        at com.orientechnologies.orient.enterprise.channel.binary.OChannelBinary.readByte(OChannelBinary.java:73) ~[orientdb-enterprise-2.1.7.jar:2.1.7]
        at com.orientechnologies.orient.enterprise.channel.binary.OChannelBinaryAsynchClient.beginResponse(OChannelBinaryAsynchClient.java:200) ~[orientdb-enterprise-2.1.7.jar:2.1.7]
        at com.orientechnologies.orient.enterprise.channel.binary.OChannelBinaryAsynchClient.beginResponse(OChannelBinaryAsynchClient.java:171) ~[orientdb-enterprise-2.1.7.jar:2.1.7]
        at com.orientechnologies.orient.client.remote.OStorageRemote.beginResponse(OStorageRemote.java:2155) ~[orientdb-client-2.1.7.jar:2.1.7]
        at com.orientechnologies.orient.client.remote.OStorageRemote.command(OStorageRemote.java:1188) ~[orientdb-client-2.1.7.jar:2.1.7]
        ... 52 common frames omitted

seems connections to server reached max size(client.channel.maxPool=100) leading to connection lose, after reduce db.pool.max  the error disappear.  will follow up.

what might be the root cause of this issue, thanks.

Seems default 100 of client.channel.maxPool is not enough in our case. any suggestion?

### Runninng Mode
- [ ] Remote

### Misc
- [ ] I have a distributed setup with multiple servers. How many?  2

### OrientDB Version
2.1.16

### Operating System
- [ ] Linux

### Java Version
- [ ] 7

**Comment by NLPScott** [comment]
me too, who can give some advise?? thanks

**Comment by tglman** [comment]
Hi all,

this log is not an error it just show that some sessions were invalidate and the client need to reconnect, it start to show now just because we increased the log level of some exception, so no action is needed, i'll try to filter in a better way the log of exception to avoid to log not needed errors.

**Comment by ray3888** [comment]
thanks @tglman , but seems it related to these error appear at times: (just guess)

> 
error=Error on executing command: sql.select from v where pk='150001984621'
Caused by: java.io.EOFException: null

**Comment by ndelaforge** [comment]
Hi all,

I have the same exception in my server log and I am also experimenting strange behaviors in production environment. I am testing OrientDB with Gremlin and the [gremlin scala](https://github.com/mpollmeier/gremlin-scala) library with doesn't support transactions (OrientGraph noTx), so I worry a lot about information loss.
For now, I keep my connection handler in a variable that is shared across my services and it is closed only on application shutdown.
Does this exception mean that the connection should be refreshed by the client each time an operation is performed on the graph instead of keeping it alive ?

Thanks in advance for you enlightments.

**Comment by ray3888** [comment]
@ndelaforge , to my knowledge it's better to use connection pool (official recommand), 
After several days' observation now I am almost sure that error "Found unknown session xx" is related to connection pool size, if network channels between client and server exceeds client.channel.maxPool existing connections may drop somehow, which may also cause occasional read error.

**Comment by ndelaforge** [comment]
Yes, I am using the connection pool. This is the way I manage my connection to orient : 
```
  lazy val factory: OrientGraphFactory = new OrientGraphFactory(uri, user, passwd).setupPool(10)
  lazy val graph: OrientGraph = factory.getNoTx(true, true)
```
Graph instance is stored in a variable.
Is orient session bound to graph instances ?
My client app is a server, is it possible that orientdb is closing sessions after some time ?

**Comment by tglman** [comment]
hi @ndelaforge,

The session is bound to the OrientGraph instance, and you should not use and OrientGraph instance across threads, each thread should have it's own OrientGraph instance got from the factory and released to it.

we usually suggest to keep the factory as global reference on the application, and each single operation acquire and release the graph instance from the factory.

speaking java:
```
OrientGraph graph =factory.getNoTx(true,true)
try {
//Do staff on the graph
} finally {
    graph.shutdown();
}
```

I didn't really understood why you are using NoTx is it there any requirement for that ?

**Comment by ndelaforge** [comment]
Thanks @tglman for you reply.
I followed your advice and I change my code according to the acquire/release strategy.
Now I'm facing issues with my client side hook.
Should I register it for each graph instance ?

**Comment by tglman** [comment]
hi @ndelaforge,

yes, hooks need to be registered one per instance.

**Comment by ray3888** [comment]
some new observation, noticed that may need to reload graphFactory to avoid occasional read error and "Found unknown session", since server may drop connections after client running for a while somehow. this may be a bug. the reload code like:
> factory.close();
> factory = new OrientGraphFactory("url", "user","password");

@tglman  can I run above code, is there any issue? thanks

**Comment by tglman** [comment]
hi @ray3888,

close the factory cause that all the graph took from the factory will be shoutdown, if that is ok for your context should not be a problem, anyway i see that as an overkill, the  "Found unknown session" is not a critical bug, and it will force a reconnection internally, that is what you will do closing and reopening the factory, but the atomatic reconnect will be limited only to the session that have problems, and the factory close will inpact everything, in any case are you using the last version? we fixed a few issue around that message.

**Comment by ray3888** [comment]
thanks @tglman  the version is 2.1.16

**Comment by tglman** [comment]
hi @ray3888,

Actually that message was fixed in 2.1.18, so i'll suggest to upgrate to 2.1.19 or to the last 2.2.x (if you have time for a bit of migration)

so if you don't mind i'll close this.

**Comment by ray3888** [comment]
@tglman  sure, thanks.
